File: frontend/src/app/services/rag.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { RagResponse } from '../models/content.model';

@Injectable({
  providedIn: 'root'
})
export class RagService {
  private apiUrl = environment.vertexAIApiUrl;
  
  constructor(private http: HttpClient) { }
  
  /**
   * Ask a question and get an answer using RAG (Retrieval-Augmented Generation)
   * @param question The question to ask
   * @param contentIds Optional array of content IDs to limit the search scope
   * @param filters Optional filters to apply to the search (tracks, tags, etc.)
   */
  askQuestion(question: string, contentIds?: string[], filters?: any): Observable<RagResponse> {
    const payload = {
      question,
      contentIds,
      filters
    };
    
    return this.http.post<RagResponse>(`${this.apiUrl}/ask`, payload).pipe(
      catchError(error => {
        console.error('Error asking question using RAG:', error);
        return throwError(() => new Error('Failed to get answer. Please try again.'));
      })
    );
  }
  
  /**
   * Summarize a document using Vertex AI
   * @param contentId ID of the content to summarize
   */
  summarizeDocument(contentId: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/summarize`, { contentId }).pipe(
      catchError(error => {
        console.error('Error summarizing document:', error);
        return throwError(() => new Error('Failed to summarize document. Please try again.'));
      })
    );
  }
  
  /**
   * Generate tags for a document using Vertex AI
   * @param contentId ID of the content to generate tags for
   */
  generateTags(contentId: string): Observable<string[]> {
    return this.http.post<string[]>(`${this.apiUrl}/generate-tags`, { contentId }).pipe(
      catchError(error => {
        console.error('Error generating tags:', error);
        return throwError(() => new Error('Failed to generate tags. Please try again.'));
      })
    );
  }
  
  /**
   * Retrieve similar documents based on a query or document ID
   * @param query Text query to find similar documents
   * @param contentId Optional content ID to find similar to this document
   * @param limit Maximum number of results to return
   */
  findSimilarDocuments(query?: string, contentId?: string, limit: number = 5): Observable<any[]> {
    const payload = {
      query,
      contentId,
      limit
    };
    
    return this.http.post<any[]>(`${this.apiUrl}/similar`, payload).pipe(
      catchError(error => {
        console.error('Error finding similar documents:', error);
        return throwError(() => new Error('Failed to find similar documents. Please try again.'));
      })
    );
  }
} 

File: backend/app/services/rag_service.py
"""
RAG (Retrieval-Augmented Generation) service for the application.
"""
import logging
from typing import Any, Dict, List, Optional

from app.core.config import settings
from app.models.content import ContentInDB

# Setup logging
logger = logging.getLogger(__name__)


class RAGService:
    """Service for Retrieval-Augmented Generation capabilities."""

    def __init__(self):
        """Initialize the RAG service."""
        self.vertex_project = settings.VERTEX_AI_PROJECT
        self.vertex_location = settings.VERTEX_AI_LOCATION
        self.model_id = settings.VERTEX_MODEL_ID

        # In production, this would initialize the Vertex AI client
        # For now, we'll use a mock implementation
        logger.info(f"Initialized RAG service with model {self.model_id}")

    def ask_question(
        self,
        question: str,
        content_ids: Optional[List[str]] = None,
        content_items: Optional[List[ContentInDB]] = None,
    ) -> Dict[str, Any]:
        """Ask a question about content.

        Args:
            question: The question to ask.
            content_ids: List of content IDs to search in (optional).
            content_items: List of content items to search in (optional).

        Returns:
            Dictionary with answer and source information.
        """
        logger.info(f"Processing question: {question}")

        # In production, this would use Vertex AI and vector search
        # For now, we'll return a mock response

        # Mock answer
        answer = f"This is a mock answer to the question: {question}"

        # Mock sources
        sources = []
        if content_items:
            for item in content_items[:2]:  # Limit to 2 sources for the mock
                sources.append(
                    {
                        "id": item.id,
                        "title": item.title,
                        "page_number": "1",  # Mock page number
                        "text_snippet": f"Relevant text from {item.title}...",  # Mock snippet
                    }
                )

        return {"answer": answer, "sources": sources, "model_used": self.model_id}

    def summarize_content(self, content: ContentInDB) -> str:
        """Generate a summary of content.

        Args:
            content: Content to summarize.

        Returns:
            Generated summary.
        """
        logger.info(f"Generating summary for content {content.id}")

        # In production, this would use Vertex AI to generate a summary
        # For now, we'll return a mock summary
        return f"This is a mock summary of the content titled '{content.title}'."

    def generate_tags(self, content: ContentInDB) -> List[str]:
        """Generate tags for content.

        Args:
            content: Content to generate tags for.

        Returns:
            List of generated tags.
        """
        logger.info(f"Generating tags for content {content.id}")

        # In production, this would use Vertex AI to generate tags
        # For now, we'll return mock tags
        return ["conference", "presentation", "sample-tag"]

    def find_similar_content(self, content_id: str, limit: int = 5) -> List[Dict[str, Any]]:
        """Find content similar to the given content.

        Args:
            content_id: ID of the content to find similar items for.
            limit: Maximum number of similar items to return.

        Returns:
            List of similar content items with similarity scores.
        """
        logger.info(f"Finding content similar to {content_id}")

        # In production, this would use vector search
        # For now, we'll return mock similar content
        return [
            {"id": "mock-similar-1", "title": "Similar Content 1", "similarity_score": 0.95},
            {"id": "mock-similar-2", "title": "Similar Content 2", "similarity_score": 0.85},
        ]
